controllers/apidefinition_controller.go

```diff
@@ -748,6 +707,211 @@ func (r *ApiDefinitionReconciler) updateStatus(
 	return r.Status().Update(ctx, &api)
 }
 
+// breakSubgraphLink breaks the link between given ApiDefinition and Subgraph object it refers to. If pass is specified,
+// the function skips updating given ApiDefinition's status.
+func (r *ApiDefinitionReconciler) breakSubgraphLink(
+	ctx context.Context,
+	desired *tykv1alpha1.ApiDefinition,
+	pass bool,
+) error {
+	if desired.Status.LinkedSubgraphName == "" {
+		return nil
+	}
+
+	subgraph := &tykv1alpha1.SubGraph{}
+	subgraphNamespacedName := types.NamespacedName{
+		Namespace: desired.ObjectMeta.Namespace,
+		Name:      desired.Status.LinkedSubgraphName,
+	}
+
+	// If ApiDefinition is linked to Subgraph (through .status.subgraph_name field) even though it's
+	// GraphRef is empty, remove the link between Subgraph CR and ApiDefinition CR respectively.
+	err := r.Client.Get(ctx, subgraphNamespacedName, subgraph)
+	if err != nil {
+		r.Log.Error(err,
+			"failed to break the link between ApiDefinition and Subgraph CRs",
+			"couldn't find Subgraph", subgraphNamespacedName.String(),
+		)
+
+		return err
+	}
+
+	if subgraph.Status.LinkedApiDefID != "" {
+		subgraph.Status.LinkedApiDefID = ""
+
+		err = r.Status().Update(ctx, subgraph)
+		if err != nil {
+			r.Log.Error(err,
+				"failed to update Subgraph status after removing ApiDefinition link",
+				"Subgraph CR", subgraphNamespacedName.String(),
+			)
+
+			return err
+		}
+	}
+
+	if !pass && desired.Status.LinkedSubgraphName != "" {
```

Comment:
Do we need a check for `desired.Status.LinkedSubgraphName` here again?
We are already checking at the start.